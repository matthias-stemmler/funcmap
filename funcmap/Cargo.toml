[package]
name = "funcmap"
version = "0.1.0"
authors = ["Matthias Stemmler <matthias.stemmler@gmail.com>"]
edition = "2021"
rust-version = "1.56" # should be the same as in Cargo.toml of funcmap_derive, docs and MSRV job
description = "Derivable functorial mappings for Rust"
readme = "crates-io.md"
repository = "https://github.com/matthias-stemmler/funcmap"
license = "MIT OR Apache-2.0"
keywords = ["data", "derive", "macro", "map", "no_std"]
categories = ["algorithms", "data-structures", "development-tools", "no-std", "rust-patterns"]
include = ["examples/**/*", "src/**/*", "Cargo.toml", "LICENSE-APACHE", "LICENSE-MIT", "crates-io.md"]

[features]
default = ["std"]
alloc = []
std = ["alloc"]

[dependencies]
"funcmap_derive" = { version = "=0.1.0", path = "../funcmap_derive" }
