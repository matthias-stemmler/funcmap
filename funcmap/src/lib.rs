#![cfg_attr(not(feature = "std"), no_std)]
#![deny(warnings)]
#![deny(missing_copy_implementations)]
#![deny(missing_debug_implementations)]
// #![deny(missing_docs)] // TODO uncomment
#![deny(unreachable_pub)]
#![deny(unused_crate_dependencies)]
#![deny(unused_extern_crates)]
#![deny(unused_lifetimes)]
#![deny(unused_qualifications)]
#![deny(clippy::all)]
#![deny(clippy::pedantic)]
#![deny(clippy::cargo_common_metadata)]
#![deny(clippy::multiple_crate_versions)]
#![deny(clippy::rest_pat_in_fully_bound_structs)]
#![deny(clippy::use_debug)]
#![allow(clippy::module_name_repetitions)]
#![allow(clippy::missing_errors_doc)] // TODO remove

mod array;
mod impls_core;

#[cfg(feature = "alloc")]
mod impls_alloc;

#[cfg(feature = "std")]
mod impls_std;

use core::{convert::Infallible, hint};

#[doc(hidden)]
pub use funcmap_derive::*;

#[derive(Copy, Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct TypeParam<const N: usize>;

pub trait FuncMap<A, B, P = TypeParam<0>>: Sized {
    type Output;

    fn try_func_map<F, E>(self, f: F) -> Result<Self::Output, E>
    where
        F: FnMut(A) -> Result<B, E>;

    fn try_func_map_over<F, E>(self, _: P, f: F) -> Result<Self::Output, E>
    where
        F: FnMut(A) -> Result<B, E>,
    {
        self.try_func_map(f)
    }

    fn func_map<F>(self, mut f: F) -> Self::Output
    where
        F: FnMut(A) -> B,
    {
        let result: Result<_, Infallible> = self.try_func_map(|value| Ok(f(value)));

        match result {
            Ok(value) => value,

            // SAFETY: `_err` is of type `Infallible`, of which no values exist, so this arm is never reached
            Err(_err) => unsafe { hint::unreachable_unchecked() },
        }
    }

    fn func_map_over<F>(self, _: P, f: F) -> Self::Output
    where
        F: FnMut(A) -> B,
    {
        self.func_map(f)
    }
}
