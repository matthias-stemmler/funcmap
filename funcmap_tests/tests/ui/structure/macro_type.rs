use funcmap::FuncMap;

macro_rules! test_type {
    () => {
        T
    };
}

#[derive(FuncMap)]
struct Test<T>(test_type!());

// try to derive `Default` to make sure the error message is the same
#[derive(Default)]
struct TestDefault<T>(test_type!());

fn main() {}
