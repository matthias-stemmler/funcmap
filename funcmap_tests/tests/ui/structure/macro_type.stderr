error: `derive` cannot be used on items with type macros
  --> tests/ui/structure/macro_type.rs:10:16
   |
10 | struct Test<T>(test_type!());
   |                ^^^^^^^^^^^^

error: `derive` cannot be used on items with type macros
  --> tests/ui/structure/macro_type.rs:14:23
   |
14 | struct TestDefault<T>(test_type!());
   |                       ^^^^^^^^^^^^
